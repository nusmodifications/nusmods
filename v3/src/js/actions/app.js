// @flow

import type { FSA } from 'types/redux';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}

export const TOGGLE_FEEDBACK_MODAL = 'TOGGLE_FEEDBACK_MODAL';
export function toggleFeedback(): FSA {
  return {
    type: TOGGLE_FEEDBACK_MODAL,
    payload: null,
  };
}
