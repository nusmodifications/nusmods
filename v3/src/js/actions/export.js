// @flow
import domtoimage from 'dom-to-image';
import { iCalForTimetable } from 'utils/ical';

export const DOWNLOAD_AS_JPEG = 'DOWNLOAD_AS_JPEG';
export function downloadAsJpeg(domElement: Element) {
  return (dispatch: Function) => {
    dispatch({
      type: `${DOWNLOAD_AS_JPEG}_PENDING`,
    });

    const style = { margin: '0', marginLeft: '-0.25em' };
    return domtoimage.toJpeg(domElement, { bgcolor: '#fff', style })
      .then((dataUrl) => {
        const link = document.createElement('a');
        link.download = 'timetable.jpeg';
        link.href = dataUrl;
        link.click();
        dispatch({
          type: `${DOWNLOAD_AS_JPEG}_SUCCESS`,
        });
      })
      .catch((err) => {
        console.error(err);
        dispatch({
          type: `${DOWNLOAD_AS_JPEG}_FAILURE`,
        });
      });
  };
}

export const DOWNLOAD_AS_ICAL = 'DOWNLOAD_AS_ICAL';
export function downloadAsIcal(semester, timetable, moduleData) {
  const results = iCalForTimetable(semester, timetable, moduleData);
  return {
    type: DOWNLOAD_AS_ICAL,
    payload: results,
  };
}
