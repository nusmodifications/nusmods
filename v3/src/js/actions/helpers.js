// @flow

import _ from 'lodash';

import type { FSA } from 'types/redux';

export const RESET_REQUEST_STATE: string = 'RESET_REQUEST_STATE';
export function resetRequestState(domain: string | Array<string>): FSA {
  return {
    type: RESET_REQUEST_STATE,
    payload: _.isArray(domain) ? domain : [domain],
  };
}

export const RESET_ERROR_STATE: string = 'RESET_ERROR_STATE';
export function resetErrorState(domain: string | Array<string>): FSA {
  return {
    type: RESET_ERROR_STATE,
    payload: _.isArray(domain) ? domain : [domain],
  };
}

export const RESET_ALL_STATE: string = 'RESET_ALL_STATE';
export function resetAllState(): FSA {
  return {
    type: RESET_ALL_STATE,
  };
}
