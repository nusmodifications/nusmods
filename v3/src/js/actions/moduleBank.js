// @flow
import type {
  ModuleCode,
} from 'types/modules';
import type { FSA } from 'types/redux';

import { API_REQUEST } from 'middlewares/requests-middleware';
import NUSModsApi from 'apis/nusmods';

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList(): FSA {
  return {
    type: FETCH_MODULE_LIST,
    payload: {
      method: 'GET',
      url: NUSModsApi.moduleListUrl(),
    },
    meta: {
      [API_REQUEST]: true,
    },
  };
}

export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModule(moduleCode: ModuleCode): FSA {
  return {
    type: FETCH_MODULE,
    payload: {
      method: 'GET',
      url: NUSModsApi.moduleDetailsUrl(moduleCode),
    },
    meta: {
      [API_REQUEST]: true,
    },
  };
}
