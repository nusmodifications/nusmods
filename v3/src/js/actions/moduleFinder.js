// @flow

import type { FSA } from 'types/redux';

export const SEARCH_MODULES = 'SEARCH_MODULES';
export function searchModules(searchTerm: string): FSA {
  return {
    type: SEARCH_MODULES,
    payload: { searchTerm },
  };
}

export const RESET_MODULE_FINDER = 'RESET_MODULE_FINDER';
export function resetModuleFinder(): FSA {
  return {
    type: RESET_MODULE_FINDER,
    payload: null,
  };
}
