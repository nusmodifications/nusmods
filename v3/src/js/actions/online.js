// @flow

import type { FSA } from 'types/redux';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}
