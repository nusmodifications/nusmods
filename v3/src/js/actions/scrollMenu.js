// @flow

import { kebabCase } from 'lodash';

import type { FSA } from 'types/redux';
import type { ScrollMenuId, ScrollMenuItem, ScrollMenuItemId } from 'types/reducers';

export function createMenuItem(label: string, id?: string): ScrollMenuItem {
  return {
    label,
    id: id || kebabCase(label),
  };
}

export const ADD_MENU_ITEM = 'ADD_MENU_ITEM';
export function addMenuItem(menuId: ScrollMenuId, itemOrLabel: string | ScrollMenuItem): FSA {
  const item: ScrollMenuItem = typeof itemOrLabel === 'string' ? createMenuItem(itemOrLabel) : itemOrLabel;

  return {
    type: ADD_MENU_ITEM,
    payload: { item, menuId },
  };
}

export const CLEAR_MENU_STATE = 'CLEAR_MENU_STATE';
export function clearMenuState(menuId: ScrollMenuId): FSA {
  return {
    type: CLEAR_MENU_STATE,
    payload: { menuId },
  };
}

export const NEXT_MENU_ITEM = 'NEXT_MENU_ITEM';
export function nextMenuItem(menuId: ScrollMenuId, after: ScrollMenuItemId): FSA {
  return {
    type: NEXT_MENU_ITEM,
    payload: { menuId, after },
  };
}

export const SET_MENU_ITEM = 'SET_MENU_ITEM';
export function setMenuItem(menuId: ScrollMenuId, itemId: ScrollMenuItemId): FSA {
  return {
    type: SET_MENU_ITEM,
    payload: { menuId, itemId },
  };
}
