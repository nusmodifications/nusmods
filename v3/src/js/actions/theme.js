// @flow
import type { FSA } from 'types/redux';
import type { ModuleCode } from 'types/modules';
import type { ColorIndex } from 'types/reducers';

export const SELECT_THEME: string = 'SELECT_THEME';
export function selectTheme(theme: string): FSA {
  return {
    type: SELECT_THEME,
    payload: theme,
  };
}

export const MODIFY_MODULE_COLOR: string = 'MODIFY_MODULE_COLOR';
export function modifyModuleColor(moduleCode: ModuleCode): FSA {
  return {
    type: MODIFY_MODULE_COLOR,
    payload: {
      activeModule: moduleCode,
    },
  };
}

export const CANCEL_MODIFY_MODULE_COLOR: string = 'CANCEL_MODIFY_MODULE_COLOR';
export function cancelModifyModuleColor(): FSA {
  return {
    type: CANCEL_MODIFY_MODULE_COLOR,
  };
}

export const SELECT_MODULE_COLOR: string = 'SELECT_MODULE_COLOR';
export function selectModuleColor(moduleCode: ModuleCode, colorIndex: ColorIndex): FSA {
  return {
    type: SELECT_MODULE_COLOR,
    payload: {
      moduleCode,
      colorIndex,
    },
  };
}

export const TOGGLE_TIMETABLE_ORIENTATION: string = 'TOGGLE_TIMETABLE_ORIENTATION';
export function toggleTimetableOrientation(): FSA {
  return {
    type: TOGGLE_TIMETABLE_ORIENTATION,
  };
}
