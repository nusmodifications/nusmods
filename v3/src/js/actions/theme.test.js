// @flow
import * as actions from 'actions/theme';

describe('theme', () => {
  test('should dispatch a select of theme', () => {
    const theme: string = 'test';
    expect(actions.selectTheme(theme)).toMatchSnapshot();
  });

  test('should dispatch a cycle of theme', () => {
    expect(actions.cycleTheme(1)).toMatchSnapshot();
    expect(actions.cycleTheme(-1)).toMatchSnapshot();
  });

  test('should dispatch a cycle of select of module color', () => {
    expect(actions.selectModuleColor('CS1010S', 0)).toMatchSnapshot();
    expect(actions.selectModuleColor('CS3216', 1)).toMatchSnapshot();
  });

  test('should dispatch a toggle of timetable orientation', () => {
    expect(actions.toggleTimetableOrientation()).toMatchSnapshot();
  });
});
