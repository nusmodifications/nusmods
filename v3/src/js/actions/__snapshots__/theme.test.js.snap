// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`theme should dispatch a cycle of select of module color 1`] = `
Object {
  "payload": Object {
    "colorIndex": 0,
    "moduleCode": "CS1010S",
  },
  "type": "SELECT_MODULE_COLOR",
}
`;

exports[`theme should dispatch a cycle of select of module color 2`] = `
Object {
  "payload": Object {
    "colorIndex": 1,
    "moduleCode": "CS3216",
  },
  "type": "SELECT_MODULE_COLOR",
}
`;

exports[`theme should dispatch a cycle of theme 1`] = `
Object {
  "payload": 1,
  "type": "CYCLE_THEME",
}
`;

exports[`theme should dispatch a cycle of theme 2`] = `
Object {
  "payload": -1,
  "type": "CYCLE_THEME",
}
`;

exports[`theme should dispatch a select of theme 1`] = `
Object {
  "payload": "test",
  "type": "SELECT_THEME",
}
`;

exports[`theme should dispatch a toggle of timetable orientation 1`] = `
Object {
  "payload": null,
  "type": "TOGGLE_TIMETABLE_ORIENTATION",
}
`;
